RT_UPDATE/download_asfr.R

```diff
@@ -0,0 +1,363 @@
+#' @param AsfrDatesIn Numeric vector of reference years for which ASFR data are
+#'   to be downloaded or aligned.
+#' @param method Character specifying interpolation method to use when ASFR data
+#'   are interpolated across years. Defaults to \code{"linear"}.
+#' @param output Character specifying desired output age structure:
+#'   \code{"single"} (single-year ages) or \code{"5-year"} (default).
+#' @param ... Additional arguments passed to the interpolation routine.
+#'
+#' @details
+#' The function expects WPP single-age percentage ASFR data (\code{percentASFR1dt})
+#' and total fertility rate (\code{tfr1}) to be available in the latest installed
+#' \code{wpp} package. For the \code{"single"} option, the function constructs
+#' single-year ASFR by multiplying percentage ASFR by TFR and interpolating to
+#' requested years. For the \code{"5-year"} option, the function first aggregates
+#' percentages to 5-year groups and then multiplies by TFR; it then interpolates
+#' group-level rates to requested years. When \code{Asfrmat} is provided by the
+#' user, it is converted to a matrix and rownames are set according to the
+#' requested \code{output}.
+#'
+#' @return
+#' A numeric matrix of ASFR values by age (rows) and year (columns).
+#'
+#' @importFrom dplyr filter select mutate group_by summarise ungroup left_join
+#' @importFrom dplyr group_nest
+#' @importFrom tidyr pivot_longer pivot_wider unnest
+#' @importFrom purrr map
+#' @importFrom tibble as_tibble column_to_rownames
+#' @importFrom readr parse_number
+#' @importFrom rlang .data
+#' #importFrom magrittr %>% 
+#'
+#' @examples
+#' \dontrun{
+#' # Download 5-year ASFR for France for three years
+#' asfr_data <- download_Asfr(location = "France",
+#'                            AsfrDatesIn = c(2000, 2005, 2010),
+#'                            output = "5-year")
+#'
+#' # Use user-supplied ASFR matrix (rows = ages, cols = years)
+#' Asfrmat <- matrix(runif(25), nrow = 5)
+#' download_Asfr(Asfrmat = Asfrmat,
+#'               AsfrDatesIn = c(2000, 2005, 2010, 2015, 2020),
+#'               output = "single")
+#' }
+#'
+#' @export
+
+
+download_Asfr <- function(Asfrmat     = NULL,
+                          location    = NULL,
+                          AsfrDatesIn = NULL,
+                          method      = "linear",
+                          output      = "5-year",
+                          ...) {
+  
+  verbose <- getOption("basepop_verbose", TRUE)
+  
+  # -------------------------------
+  # CASE 1: user-supplied Asfrmat
+  # -------------------------------
+  # If the user provides Asfrmat we simply coerce to matrix, assign column names
+  # (years) and set rownames (ages) according to requested output format.
+  #
+  # NOTE: Asfrmat is expected to be in conventional reproductive ages
+  # (usually 15-49 single-year ages or 5-year groups 15-19,...). If input
+  # has a different age basis, it should be transformed first.
+  if(!is.null(Asfrmat)) {
```

Comment:
this is a fragile block. if assigning dimnames, first check lengths match, but also check that it's needed, and indeed the dates-in information might already be in the colnames, which means that's an optional arg too.